### PR TITLE
[DS-186][DS-192] Analysis of MR and Preprocessor Scores 

### PR DESCRIPTION
**Pull request by happilyeverafter95**
*Goals:*

The goal of this PR is to better understand the Mnemonic Reader and preprocessor scores and whether they give any indication on the accuracy of Mnemonic Reader's answer. 

*Notes for Reviewers:*

Files used for analysis can be found in Google drives: Everyone > Data Science > Test Data

Short write up for the analysis can be found in Data Science confluence (Title: Analysis of Confidence Scores for Mnemonic Reader) 

*Changes Proposed in this PR:*

- Added script to analyze Mnemonic Reader and Preprocessor scores
- Script creates plot of false positives and false negatives for human answerable questions based on preprocessor scores
- Script also plots answer accuracy based on Mnemonic Reader's confidence scores

### CODE REVIEW

File: analysis of confidence scores.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+answerable = data[data['Answerable']==1]
+
+# Not Human Answerable questions 
+notAnswerable = data[data['Answerable']==0]
+
+# Confidence Score plot between the two categories 
+plt.hist(answerable['Cos Sim'], histtype='stepfilled', color= "blue", alpha= 0.3, label = 'Answer')
+plt.hist(notAnswerable['Cos Sim'], histtype='stepfilled', color="red", alpha= 0.3, label = 'Do not answer')
+
+plt.xlabel("Confidence scores")
```
Comment (Labibahx) [inline analysis of confidence scores.py]: these are plotted as confidence scores but the are tabulated as cos sim. Can you please clarify what this quantity is?

Comment (happilyeverafter95) [reply analysis of confidence scores.py]: Cosine similarity is being used as the confidence score.

File: analysis of confidence scores.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+## <---------------- Preprocessor Accuracy vs Score --------------------> 
+
+# Filter for human ansewrable Questions
+dataAnswer = data[data['Answerable'] ==1]
+
+# Considering Top 5 Accuracies 
+topFiveAcc = dataAnswer[dataAnswer['Acc Top 5'] == True]
+topFiveInacc = dataAnswer[dataAnswer['Acc Top 5'] == False]
+
```
Comment (Labibahx) [inline analysis of confidence scores.py]: Acc Top 5 indicates the correct match is in the top 5 returned answers?

Comment (happilyeverafter95) [reply analysis of confidence scores.py]: yes :)

File: analysis of confidence scores.py (hunk 1 above)
Comment (Labibahx) [reply analysis of confidence scores.py]: Have you checked how robust the cos similarity in capturing the confidence? Is it always successful in representing whether an answer is the correct match or not?

Comment (happilyeverafter95) [reply analysis of confidence scores.py]: I have some results here: https://kiitehq.atlassian.net/wiki/spaces/PROD/pages/270893057/Analysis+of+Confidence+Scores+for+Mnemonic+Reader?focusedCommentId=270958624&src=mail&src.mail.timestamp=1540596862809&src.mail.notification=com.atlassian.confluence.plugins.confluence-notifications-batch-plugin%3Abatching-notification&src.mail.recipient=8a7f8085652fc631016533814982008b&src.mail.action=view#comment-270958624

In a nutshell, its good but not great.

Comment (happilyeverafter95) [reply analysis of confidence scores.py]: It does. This part of the script is comparing whether a question is answerable by a human to the preprocessor's score. For this aspect, the preprocessor's score is more reliable than Mnemonic Reader's score.

Comment (happilyeverafter95) [reply analysis of confidence scores.py]: Sorry, I just realized that my previous response might be confusing. 

For clarity, the column labels:  

"cos sim" = preprocessor's confidence score which is also the cosine similarity 
"Confidence" = Mnemonic Reader's confidence score
